src/main/scala-2.13+/scala/concurrent/stm/skel/AtomicArray.scala

```diff
@@ -0,0 +1,279 @@
+  }
+
+  override def className = "AtomicArray"
+
+  /** Clones this object, including the underlying Array. */
+  override def clone: AtomicArray[T] = {
+    val b = newSpecificBuilder
+    b.sizeHint(length)
+    b ++= this
+    b.result()
+  }
+
+  // this should not be abstract: https://github.com/scala/bug/issues/11006
+  def mapInPlace(f: T => T): this.type = {
+    var i = 0
+    while (i < length) {
+      update(i, f(apply(i)))
+      i = i + 1
+    }
+    this
+  }
+}
+
+object AtomicArray extends StrictOptimizedClassTagSeqFactory[AtomicArray] {
+
+  val untagged: SeqFactory[AtomicArray] = new ClassTagSeqFactory.AnySeqDelegate(this)
+
+  def empty[T](implicit m: ClassTag[T]): AtomicArray[T] =
+    AtomicArray[T](0)
+
+  def from[T](it: IterableOnce[T])(implicit m: ClassTag[T]): AtomicArray[T] =
+    AtomicArray[T](it)
+
+  def newBuilder[T](implicit m: ClassTag[T]): mutable.Builder[T, AtomicArray[T]] =
+    AtomicArrayBuilder.of(m)
+
+  def apply[T](size: Int)(implicit m: ClassTag[T]): AtomicArray[T] = {
+    (m.newArray(0).asInstanceOf[AnyRef] match {
+      case _: Array[Boolean] => new ofBoolean(size)
+      case _: Array[Byte]    => new ofByte(size)
+      case _: Array[Short]   => new ofShort(size)
+      case _: Array[Char]    => new ofChar(size)
+      case _: Array[Int]     => new ofInt(size)
+      case _: Array[Float]   => new ofFloat(size)
+      case _: Array[Long]    => new ofLong(size)
+      case _: Array[Double]  => new ofDouble(size)
+      case _: Array[Unit]    => new ofUnit(size)
+      case _: Array[AnyRef]  => new ofRef[AnyRef](size)
+    }).asInstanceOf[AtomicArray[T]]
+  }
+
+  def apply(elems: Array[Boolean]) = new ofBoolean(new AtomicIntegerArray(elems map {if(_) 1 else 0}))
+  def apply(elems: Array[Byte])    = new ofByte(   new AtomicIntegerArray(elems map {_.toInt}))
+  def apply(elems: Array[Short])   = new ofShort(  new AtomicIntegerArray(elems map {_.toInt}))
+  def apply(elems: Array[Char])    = new ofChar(   new AtomicIntegerArray(elems map {_.toInt}))
+  def apply(elems: Array[Int])     = new ofInt(    new AtomicIntegerArray(elems))
+  def apply(elems: Array[Float])   = new ofFloat(  new AtomicIntegerArray(elems map {java.lang.Float.floatToRawIntBits(_)}))
+  def apply(elems: Array[Long])    = new ofLong(   new AtomicLongArray(elems))
+  def apply(elems: Array[Double])  = new ofDouble( new AtomicLongArray(elems map {java.lang.Double.doubleToRawLongBits(_)}))
+  def apply(elems: Array[Unit])    = new ofUnit(   elems.length)
+  def apply[T <: AnyRef](elems: Array[T]) =
+    new ofRef(new AtomicReferenceArray(elems.asInstanceOf[Array[AnyRef]]).asInstanceOf[AtomicReferenceArray[T]])
+
+  def apply[T](elems: IterableOnce[T])(implicit m: ClassTag[T]): AtomicArray[T] = {
+    val array: Array[_] = elems match {
+      case w: mutable.WrappedArray[_] => w.array // we're going to copy out regardless, no need to duplicate right now
+      case _ => elems.iterator.toArray
```

Comment:
This seems like a semantic change to the old code which was just `elems.toArray`. Is there a reason for inserting the iterator here? I assume that `apply(Array(1, 2, 3)` creates the `WrappedArray` case, and so I wonder what case could be the catch-all where `toArray` without iterator would cause a problem?

Reply:
> Is there a reason for inserting the iterator here?

That’s the new way of calling operations on `IterableOnce` instances. That being said, we should probably add the following case to handle the situations where `toArray` on a concrete collection is more efficient than building an array from an `Iterator`:

```suggestion
      case iterable: Iterable[_] => iterable.toArray
      case _ => elems.iterator.toArray
```

Reply:
Ok, I see the deprecated implementation in `IterableOnce`. Yeah, adding this third case seems good.